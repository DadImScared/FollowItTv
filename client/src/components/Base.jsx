
import React, { Component } from 'react';

import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import cyan from 'material-ui/colors/cyan';
import deepPurple from 'material-ui/colors/deepPurple';
import { MuiThemeProvider, createMuiTheme, withStyles } from 'material-ui/styles';
import CssBaseline from 'material-ui/CssBaseline';

import Navbar from './Navbar';
import Navdrawer from './Navdrawer';
import Main from './Main';

import { getFollowedShows } from '../actions/followedShows';
import styles from '../styles/Base.css';

const theme = createMuiTheme({
  palette: {
    primary: cyan,
    secondary: deepPurple
  }
});

export class  Base extends Component {

  constructor(...args) {
    super(...args);
    this.state = {
      isOpen: false
    };
  }

  async componentDidMount() {
    try {
      await getFollowedShows(this.props.dispatch);
    }
    catch ({ response: { data, status } }) {
      if (status === 401) {
        // do nothing if users token expired or is wrong
        return;
      }
      console.log(data);
      console.log(status);
    }
  }

  toggleNavdrawer = () => {
    this.setState((prevState => ({ isOpen: !prevState.isOpen })));
  };

  render() {
    const { classes, ...other } = this.props;
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline/>
        <div className={classes.appFrame}>
          <Navbar toggleNav={this.toggleNavdrawer} />
          <Navdrawer toggleNav={this.toggleNavdrawer} isOpen={this.state.isOpen} />
          <Main {...other} />
        </div>
      </MuiThemeProvider>
    );
  }
}

export default withStyles(styles)(withRouter(connect()(Base)));
