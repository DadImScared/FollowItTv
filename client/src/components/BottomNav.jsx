
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import _ from 'lodash';
import classNames from 'classnames';

import { withStyles } from '@material-ui/core/styles';
import BottomNavigation from '@material-ui/core/BottomNavigation';
import BottomNavigationAction from '@material-ui/core/BottomNavigationAction';
import Hidden from '@material-ui/core/Hidden';

import MenuIcon from '@material-ui/icons/Menu';
import FavoriteIcon from '@material-ui/icons/Favorite';
import ClockIcon from 'mdi-material-ui/ClockOut';

import styles from '../styles/BottomNavbar';

const navActions = [
  ['schedule', ClockIcon],
  ['menu', MenuIcon],
  ['my shows', FavoriteIcon]
];

export class BottomNav extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      value: 5
    };
  }

  handleChange = (event, value) => {
    const { history, toggleNavdrawer, location: { pathname } } = this.props;
    switch (value) {
      case 0:
        if (pathname === '/schedule') return;
        history.push('/schedule');
        break;
      case 1:
        toggleNavdrawer();
        this.setState({ value: 1 });
        break;
      case 2:
        if (pathname === '/my_shows') return;
        history.push('my_shows');
        break;
      default:
        return;
    }
  };

  componentDidMount() {
    const { location: { pathname } } = this.props;
    this.setValueFromPathname(pathname);
  }

  componentWillReceiveProps(nextProps) {
    const { location: { pathname: oldPathname }, navDrawerOpen: oldDrawerOpen } = this.props;
    const { location: { pathname }, navDrawerOpen } = nextProps;
    if ((oldPathname !== pathname) || (oldDrawerOpen !== navDrawerOpen && !navDrawerOpen)) {
      this.setValueFromPathname(pathname);
    }
  }

  setValueFromPathname = (pathname) => {
    if (pathname.includes('/schedule')) {
      this.setState({ value: 0 });
    }
    else if (pathname.includes('/my_shows')) {
      this.setState({ value: 2 });
    }
    else {
      this.setState({ value: 3 }); // highlight no menu option
    }
  };

  render() {
    const { classes, bottomOfPage, directionDown } = this.props;
    return (
      <Hidden mdUp>
        <BottomNavigation
          value={this.state.value}
          className={classNames(classes.navBar, {
            [classes.hideNavBar]: !bottomOfPage && directionDown,
            [classes.showNavBar]: !directionDown
          })}
          onChange={this.handleChange}
          showLabels
        >
          {
            navActions.map(([actionName, ActionIcon], index) => (
              <BottomNavigationAction
                label={_.startCase(actionName)}
                icon={<ActionIcon />}
                key={index}
              />
            ))
          }
        </BottomNavigation>
      </Hidden>
    );
  }
}

BottomNav.propTypes = {
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  classes: PropTypes.shape({
    navBar: PropTypes.string.isRequired,
    hideNavBar: PropTypes.string.isRequired,
    showNavBar: PropTypes.string.isRequired
  }).isRequired,
  bottomOfPage: PropTypes.bool.isRequired,
  directionDown: PropTypes.bool.isRequired,
  navDrawerOpen: PropTypes.bool.isRequired
};

export default withStyles(styles)(BottomNav);
