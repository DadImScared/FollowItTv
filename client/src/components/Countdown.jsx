
import React, { Component } from 'react';

import moment from 'moment';

import Typography from 'material-ui/Typography';


export class Countdown extends Component {
  constructor(...args) {
    super(...args);
    this.state = {};
  }

  componentDidMount() {
    const { eventTime, startTime = moment(), callBack } = this.props;
    const diff = eventTime.diff(startTime);
    const duration = moment.duration(diff);
    if (duration.asMilliseconds() <= 0) {
      callBack && callBack();
      return;
    }
    this.setState({ duration });
    this.interval = setInterval(() => {
      this.updateTimer();
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  updateTimer = () => {
    const newDuration = this.state.duration.clone();
    newDuration.subtract(1, 'seconds');
    if (newDuration.asMilliseconds() <= 0) {
      const { callBack } = this.props;
      clearInterval(this.interval);
      callBack && callBack();
      return;
    }
    this.setState({ duration: newDuration });
  };

  getCountdown = () => {
    const { duration } = this.state;
    const days = duration.days();
    const hours = this.format(duration.hours());
    const minutes = this.format(duration.minutes());
    const seconds = this.format(duration.seconds());
    // don't show days if days are 0
    return `in ${days !== 0 ? `${days} days and `:''}${hours}:${minutes}:${seconds}`;
  };

  format = (time) => {
    const timeStr = time.toString();
    return timeStr.length === 1 ? (
      `0${timeStr}`
    ):(
      timeStr
    );
  };

  render() {
    return (
      <Typography>
        {
          this.state.duration ?
            this.getCountdown()
            :
            '00:00:'
        }
      </Typography>
    );
  }
}

export default Countdown;
