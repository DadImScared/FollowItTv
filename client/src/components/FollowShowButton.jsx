
import React, { Component } from 'react';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';

import Heart from 'mdi-material-ui/Heart';
import HeartBroken from 'mdi-material-ui/HeartBroken';

import * as followShowActions from '../actions/followedShows';
import { addShow, getShow } from '../actions/shows';
import styles from '../styles/FollowShowButton.css';


export class FollowShowButton extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      isLoading: false
    };
  }

  handleClick = async () => {
    const { followedShows, showId, shows, followShow, unfollowShow } = this.props;
    if (this.state.isLoading) {
      return;
    }
    this.setState({ isLoading: true });
    if (!shows[showId]) {
      const show = await this.getAndUpdateShows(showId);
      shows[show.id] = show;
    }
    // follow or unfollow show
    const { schedule: { days } } = shows[showId];
    try {
      await followShowActions.postFollow(showId);
      if (followedShows.includes(showId)) {
        unfollowShow(days, showId);
      }
      else {
        followShow(days, showId);
      }
    }
    catch ({ response: { data, status } }) {
      console.log(status, data);
    }
    this.setState({ isLoading: false });
  };

  getAndUpdateShows = async (showId) => {
    try {
      const { data } = await getShow(showId);
      addShow(data.id, data);
      return data;
    }
    catch ({ response: { data, status } }) {
      console.log(data, status);
    }
  };

  render() {
    const { followedShows, showId, classes } = this.props;
    return (
      <Button onClick={this.handleClick}>
        <span className={classes.heartIcon}>
          { followedShows.includes(showId) ? <HeartBroken />:<Heart /> }
        </span>
        { followedShows.includes(showId) ? 'Un follow':'Follow'}
      </Button>
    );
  }
}

const mapStateToProps = ({ shows, followedShowsById }) => ({
  shows,
  followedShows: followedShowsById
});

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({
    followShow: followShowActions.followShow,
    unfollowShow: followShowActions.unfollowShow,
    addShow
  }, dispatch);
};

export default withStyles(styles)(connect(mapStateToProps, mapDispatchToProps)(FollowShowButton));
