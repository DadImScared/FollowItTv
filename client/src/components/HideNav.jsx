
import  React from 'react';
import PropTypes from 'prop-types';

import AppBar from '@material-ui/core/AppBar';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

import styles from '../styles/HideNav.css';

const HideNav = ({ classes, directionDown, children }) => (
  <AppBar
    className={classNames(classes.appBar, {
      [classes.stickyDown]: directionDown,
      [classes.stickyUp]: !directionDown
    })}
    color={'secondary'}
    position={'sticky'}
  >
    {children}
  </AppBar>
);

HideNav.propTypes = {
  classes: PropTypes.shape({
    appBar: PropTypes.string.isRequired,
    stickyDown: PropTypes.string.isRequired,
    stickyUp: PropTypes.string.isRequired
  }).isRequired,
  directionDown: PropTypes.bool.isRequired
};

export default withStyles(styles)(HideNav);
