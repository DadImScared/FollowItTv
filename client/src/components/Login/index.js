
import React, { Component } from 'react';

import { connect } from 'react-redux';

import { logInUser } from '../../actions/users';
import withFormState from '../withFormState';
import View from './View';


const message = `
Email is not verified. Please 
`;

export class Login extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      email: '',
      shouldShow: false
    };
  }

  submit = async (e) => {
    e.preventDefault();
    this.props.submit();
    const { handleErrorResponse, form: { email, password }, history, dispatch } = this.props;
    try {
      await dispatch(logInUser({ email, password }));
      history.push('/');
    }
    catch ({ response: { data } }) {
      if (data.non_field_errors && data.non_field_errors[0] === 'E-mail is not verified.') {
        this.setState({ shouldShow: true, email });
      }
      handleErrorResponse(data);
    }
  };

  render() {
    return (
      <View {...this.state} {...this.props} submit={this.submit} message={message} />
    );
  }
}

export default withFormState(connect()(Login));
