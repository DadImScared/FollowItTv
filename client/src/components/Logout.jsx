
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

import { logOutUser } from '../actions/users';

class Logout extends Component {
  componentDidMount() {
    this.props.logOut();
  }

  render() {
    return <Redirect to='/login' />;
  }
}

const mapDispatchToProps = (dispatch) => ({
  logOut: () => dispatch(logOutUser())
});

export default connect(null, mapDispatchToProps)(Logout);
