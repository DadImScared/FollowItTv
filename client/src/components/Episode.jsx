
import React, { Component } from 'react';

import { Link } from 'react-router-dom';

import classnames from 'classnames';
import { withStyles } from 'material-ui/styles';
import Button from 'material-ui/Button';
import Card, { CardHeader, CardMedia, CardContent, CardActions } from 'material-ui/Card';
import Typography from 'material-ui/Typography';
import Collapse from 'material-ui/transitions/Collapse';
import IconButton from 'material-ui/IconButton';
import SvgIcon from 'material-ui/SvgIcon';

import Television from 'mdi-material-ui/Television';
import ExpandMoreIcon from 'material-ui-icons/ExpandMore';

import Summary from './Summary';
import FollowShowButton from './FollowShowButton';
import styles from '../styles/Episode.css';


export class Episode extends Component {
  constructor(...args) {
    super(...args);
    this.state = {
      isOpen: false,
      height: 0
    };
  }

  handleExpand = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  setRef = (el) => {
    this.el = el;
  };

  componentDidMount() {
    if (this.el) {
      this.setState({ height: this.el.clientHeight });
    }
  }

  render() {
    const {
      item: {
        name,
        show,
        airtime,
        airdate,
        summary,
        image
      } = {},
      shows = {},
      showActions = true,
      classes,
      followProps = {}
    } = this.props;
    const currentShow = this.props.show || shows[show];
    const currentId = (this.props.show && this.props.show.id) || show;
    return (
      <div>
        <Card>
          <CardHeader
            title={
              <div style={{ display: 'flex', justifyContent: 'space-between', flexWrap: 'wrap', alignItems: 'center' }}>
                <Typography variant={'subheading'}>
                  {this.props.show ? name:currentShow.name || 'no name'}
                </Typography>
                <div style={{ display: 'flex', alignItems: 'center' }}>
                  <Typography style={{ marginRight: '4px' }}>
                    {currentShow.network && currentShow.network.name}
                  </Typography>
                  <a href={`${currentShow.officialSite}`} style={{ display: 'flex' }}>
                    <SvgIcon color={'primary'}>
                      <Television />
                    </SvgIcon>
                  </a>
                </div>
              </div>
            }
            subheader={`${this.props.show ? `${airdate} | `:''}${airtime}${!this.props.show ? ` | ${name}`:''}`}
          />
          <CardMedia
            component={'img'}
            className={classes.media}
            // image={shows[show] && shows[show].image && shows[show].image.medium}
            image={(image && image.medium) || (currentShow.image && currentShow.image.medium)}
            title={`episode ${name}. show ${currentShow.name || 'no name'}`}
          />
          <CardContent>
            <Collapse collapsedHeight={'40px'} in={this.state.isOpen}>
              <CardContent>
                <Typography component={'div'}>
                  {
                    summary ?
                      <Summary summary={summary} setRef={this.setRef} />
                      :
                      'No summary entered.'
                  }
                </Typography>
              </CardContent>
            </Collapse>
            <div
              id={'read-more'}
              style={{
                display: 'flex',
                justifyContent: 'space-between',
                visibility: this.state.height >= 40 ? 'visible':'hidden'
              }}>
              <Typography style={{ alignSelf: 'center' }}>read more</Typography>
              <IconButton
                color={'secondary'}
                className={classnames(classes.expand, {
                  [classes.expandOpen]: this.state.isOpen
                })}
                onClick={this.handleExpand}
              >
                <ExpandMoreIcon />
              </IconButton>
            </div>
          </CardContent>
          {
            showActions ?
              <CardActions>
                <div style={{ display: 'flex', justifyContent: 'space-between', width: '100%' }}>
                  <FollowShowButton {...followProps} showId={currentId} />
                  <Button component={Link} to={`/show/${currentId}/general`} color={'secondary'}>
                    View show
                  </Button>
                </div>
              </CardActions>
              :
              null
          }
        </Card>
      </div>
    );
  }
}

export default withStyles(styles)(Episode);
