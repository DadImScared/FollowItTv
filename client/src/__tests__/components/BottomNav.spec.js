
import React from 'react';

import sinon from 'sinon';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { BottomNav } from '../../components/BottomNav';

describe('BottomNav', () => {
  let props;
  beforeEach(() => {
    props = {
      location: {
        pathname: '/schedule'
      },
      history: {
        push: sinon.spy()
      },
      classes: {
        navBar: 'navBar',
        hideNavBar: 'hideNavBar',
        showNavBar: 'showNavBar'
      },
      bottomOfPage: false,
      directionDown: false,
      navDrawerOpen: false,
      toggleNavdrawer: sinon.spy()
    };
  });

  it('should render', () => {
    const wrapper = shallow(<BottomNav {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('click Menu icon', () => {
    const wrapper = shallow(<BottomNav {...props} />);
    const instance = wrapper.instance();
    instance.handleChange({}, 1);
    expect(props.toggleNavdrawer.calledOnce).toEqual(true);
    expect(instance.state.value).toEqual(1);
  });

  test('click my shows icon', () => {
    const wrapper = shallow(<BottomNav {...props} />);
    const instance = wrapper.instance();
    instance.handleChange({}, 2);
    expect(props.history.push.calledOnce).toEqual(true);
  });

  test('click clock icon', () => {
    const wrapper = shallow(<BottomNav {...{ ...props, location: { pathname: '/my_shows' } }} />);
    const instance = wrapper.instance();
    instance.handleChange({}, 0);
    expect(props.history.push.calledOnce).toEqual(true);
  });

  it('should change to schedule when route pathname changes to /schedule/*', () => {
    const newProps = { ...props };
    newProps.location.pathname = '/my_shows';
    const wrapper = shallow(<BottomNav {...newProps} />);
    wrapper.setProps({ location: { pathname: '/schedule ' } });
    expect(wrapper.state()).toMatchSnapshot();
  });

  it('should change to my shows when route pathname changes to /my_shows/*', () => {
    const wrapper = shallow(<BottomNav {...props} />);
    wrapper.setProps({ location: { pathname: '/my_shows' } });
    expect(wrapper.state()).toMatchSnapshot();
  });

  it('should change to my shows when nav drawer closes and pathname is under /my_shows/*', () => {
    const wrapper = shallow(<BottomNav {...{ ...props, navDrawerOpen: true }} />);
    wrapper.setState({ value: 1 }); // menu icon clicked
    wrapper.setProps({ navDrawerOpen: false }); // menu closed
    expect(wrapper.state()).toMatchSnapshot();
  });
});
