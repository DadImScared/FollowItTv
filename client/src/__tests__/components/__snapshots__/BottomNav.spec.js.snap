// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomNav should change to my shows when nav drawer closes and pathname is under /my_shows/* 1`] = `
Object {
  "value": 0,
}
`;

exports[`BottomNav should change to my shows when route pathname changes to /my_shows/* 1`] = `
Object {
  "value": 2,
}
`;

exports[`BottomNav should change to schedule when route pathname changes to /schedule/* 1`] = `
Object {
  "value": 0,
}
`;

exports[`BottomNav should render 1`] = `
<Hidden
  implementation="js"
  lgDown={false}
  lgUp={false}
  mdDown={false}
  mdUp={true}
  smDown={false}
  smUp={false}
  xlDown={false}
  xlUp={false}
  xsDown={false}
  xsUp={false}
>
  <WithStyles(BottomNavigation)
    className="navBar showNavBar"
    onChange={[Function]}
    showLabels={true}
    value={0}
  >
    <WithStyles(BottomNavigationAction)
      icon={<Icon />}
      key="0"
      label="Schedule"
    />
    <WithStyles(BottomNavigationAction)
      icon={<pure(Menu) />}
      key="1"
      label="Menu"
    />
    <WithStyles(BottomNavigationAction)
      icon={<pure(Favorite) />}
      key="2"
      label="My Shows"
    />
  </WithStyles(BottomNavigation)>
</Hidden>
`;
